homeassistant/components/media_player/firetv.py

```diff
@@ -4,166 +4,138 @@
-                DEVICE_STATE_URL.format(
-                    self.proto, self.host, self.port, self.device_id
-                    ), timeout=10).json()
-            return response.get('state', STATE_UNKNOWN)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve device state for %s", self.device_id)
-            return STATE_UNKNOWN
 
-    @property
-    def current_app(self):
-        """Return the current app."""
-        try:
-            response = requests.get(
-                DEVICE_APPS_URL.format(
-                    self.proto, self.host, self.port, self.device_id, 'current'
-                    ), timeout=10).json()
-            _current_app = response.get('current_app')
-            if _current_app:
-                return _current_app.get('package')
-
-            return None
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve current app for %s", self.device_id)
-            return None
+def adb_decorator(override_available=False):
+    """Send an ADB command if the device is available."""
+    def adb_wrapper(func):
+        """Wait until previous ADB commands have finished."""
+        @functools.wraps(func)
+        def _adb_wrapper(self, *args, **kwargs):
+            # If the device is unavailable, don't do anything
+            if not self.available and not override_available:
+                return None
 
-    @property
-    def running_apps(self):
-        """Return a list of running apps."""
-        try:
-            response = requests.get(
-                DEVICE_APPS_URL.format(
-                    self.proto, self.host, self.port, self.device_id, 'running'
-                    ), timeout=10).json()
-            return response.get('running_apps')
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve running apps for %s", self.device_id)
-            return None
-
-    def action(self, action_id):
-        """Perform an action on the device."""
-        try:
-            requests.get(DEVICE_ACTION_URL.format(
-                self.proto, self.host, self.port, self.device_id, action_id
-                ), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Action request for %s was not accepted for device %s",
-                action_id, self.device_id)
-
-    def start_app(self, app_name):
-        """Start an app."""
-        try:
-            requests.get(DEVICE_APPS_URL.format(
-                self.proto, self.host, self.port, self.device_id,
-                app_name + '/start'), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not start %s on %s", app_name, self.device_id)
+            # Acquire the lock, try to run the command, then release the lock
+            with self.adb_lock:
```

Comment:
We should do a non blocking acquire of the lock, to avoid blocking any threads in the thread pool. Either write our own context manager to do this or use the lock methods to acquire and release the lock.

https://docs.python.org/3/library/threading.html#threading.Lock.acquire

Reply:
Isn't that what `with self.adb_lock:` does? 

(See the very bottom of the page)
https://docs.python.org/3/library/threading.html

Reply:
No, since we want to acquire non blocking.
```py
lock.acquire(blocking=True)
```

Reply:
Fixed in https://github.com/home-assistant/home-assistant/pull/17767/commits/769bb83a9cabbb7dc3563c4f38628369ce9067eb